features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,310 @@
+Each event type specifies possible mutation keys.
+- If a mutation key is absent: no mutation would be performed for that key.
+- If a mutation key is present: mutation would be performed based on the value.
+
+If an entity is mutated, subsequent events will used the mutated value.
+
+Developer is responsible for ensuring correct order of event delivery. For
+example, in most case, developer would like mutating event-handlers (e.g. 
+populating default value for some fields) before non-mutating event-handlers
+(e.g. validating field values).
+
+
+## Event Management
+
+### Alerts
+If an event delivery is permanently failed, an ERROR level log entry is
+generated to notify developers.
+
+### Past Events
+An API is provided to list past events. This API can be used to reconcile
+self-managed database with the failed events.
+
+> NOTE: BEFORE events are not persisted, regardless of success/failure.
+
+### Manual Re-delivery
+Developer can manually trigger a re-delivery of failed event, bypassing the
+retry interval limit.
 
-| key | description | require |
-| -------- | -------- | -------- | 
-| `error` | [string] An error message. | ✓ |
-| `code` | [number] An error code. |  |
+> NOTE: BEFORE events cannot be re-delivered.
 
-example:
 
+## Security
+
+Web-hook delivery process can be modeled as following:
 ```
-{
-    "code": 1001,
-    "message": "EVERYONE LOVES CAT!"
-}
+ Int. Net. ┊    Internet    ┊      Internal Network
+           ┊                ┊
+ Auth Gear─┊─────┬──────────┊>Gateway────────────>Internal Handler
+           ┊     │          └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
+           ┊     └───────────>External Handler
 ```
 
-## Configure webhook
+Considering the authenticity, integrity and confidentiality of the process:
+
+We assume `Gateway -> Internal Handler` data-path is secure with respect to
+the mentioned 3 properties; the admins of internal network (e.g. cluster admin)
+is responsible for its security.
+
+To protect the data-paths `Auth Gear -> Gateway` and
+`Auth Gear -> External Handler`, we provide some security measures:
+- Require HTTPS: ensure integrity and confidentiality of the delivery
+- Signed Request: ensure authenticity of the delivery
+
+### HTTPS
+
+We require all external web-hook handler HTTP endpoint to use HTTPS.
+
+For internal handlers, network admin is responsible to ensure the request URL
+passed to auth gear is secure.
+
+### Signature
+
+Each web-hook event request is signed with a secret key shared between auth
+gear and web-hook handler. Developer must validate the signature and reject
+requests with invalid signature to ensure the request is originated from auth
+gear.
+
+For the detail on signature generation and validation, refer to #300.
+
+> For advanced end-to-end security scenario, some network admin may wish to
+> use mTLS for authentication. We do not support this at the moment.
+
+
+# Considerations
+
+### Recursive Web-hooks
```

Comment:
Something popup in my mind. I remember in the early days when we discuss calling between gears / microservices at Skygear, we talk about adding a field in the header as a counter, to avoid recursive calling of functions (it doesn't have to be web-hooks)

Has this design been written in spec finally? (Sorry actually I don't recall if we have even designed the mechanism for a cloud functions calling another cloud functions, like how to verify the data are trusted, etc) Maybe that can be re-used here? (Since webhook is technically another way a cloud functions -- auth gear calling another cloud functions if it is inside Skygear)

Reply:
Technically speaking, the developer is free to call our endpoints by hand. In that case, they do not know they should include that counter header. We could make that header mandatory. However I can predict that every developer will make the mistake of omitting the header.

Reply:
DX-wise, we would probably need to provide a set of library/SDK (for some supported languages like Typescript, Python etc) to handle at least two things for web-hook calling / call between microservices / cloud functions:

1. How to authenticate the caller's identities (users identity or some sort of master key)
2. Avoid an infinite recursive loop of calling.

Anyway it seems beyond the scope here. I created another feature issue #342 for that. Feel free to comment there.